server/src/middleware/errorHandler.ts

```diff
@@ -0,0 +1,70 @@
+// SPDX-FileCopyrightText: 2023 The Aalto Grades Developers
+//
+// SPDX-License-Identifier: MIT
+
+import { NextFunction, Request, Response } from 'express';
+import { ValidationError } from 'yup';
+
+export enum HttpCode {
+  BadRequest = 400,
+  Unauthorized = 401,
+  NotFound = 404,
+  InternalServerError = 500,
+  BadGateway = 502
+}
+
+export class CustomError extends Error {
+  public readonly statusCode: number;
+
+  constructor(message: string, statusCode: HttpCode) {
+    super(message);
+    this.statusCode = statusCode;
+    this.name = 'CustomError';
+  }
+}
+
+export class SisuError extends CustomError {
+  public readonly sisuErrorCode: number;
+
+  constructor(sisuErrorCode: number) {
+    super(`external API error: ${sisuErrorCode}`, HttpCode.BadGateway);
+    this.sisuErrorCode = sisuErrorCode;
+  }
+}
+
+export function ErrorHandler(err: unknown, req: Request, res: Response, next: NextFunction): void {
+
+  if (err instanceof CustomError) {
+    res.status(err.statusCode);
+
+    if (err instanceof SisuError) {
+      res.send({
+        success: false,
+        errors: [err.message],
+        sisuStatusCode: err.sisuErrorCode
```

Comment:
Should this be here at all? We're already including the Sisu status in the error message, though it could be justified to include it separately as well. We could discuss this at some point.

Reply:
This was more to demonstrate how we can make even more customized errors from our "base" custom error. Unless client needs the sisu status code in number format this is useless. If we don't need to send the sisu status code separately, then we don't need `SisuError`. We could just use the `customError` in the sisu controllers as follows:
```javascript
throw new CustomError(`external API error: ${sisuStatusCode}`, HttpCode.BadGateway);
```

If we don't need the `SisuError`, I could delete it for now. Later on we can create new errors if we need properties/methods etc. that are not present in the base error.

What do you think?

Reply:
Frontend developers said that they likely wouldn't need Sisu's status code, so this error type could be deleted.

Reply:
sisu error type removed a9dbb68d1d42dc62c64a95e5554e8c19e16c405e